packages/stepper/src/Stepper.tsx

```diff
@@ -0,0 +1,319 @@
+import React, { useState } from 'react';
+import PropTypes from 'prop-types';
+import { cx } from '@leafygreen-ui/emotion';
+import CheckmarkIcon from '@leafygreen-ui/icon/dist/Checkmark';
+import EllipsisIcon from '@leafygreen-ui/icon/dist/Ellipsis';
+import {
+  completedStepLabelStyle,
+  containerStyle,
+  currentStepLabelStyle,
+  currentStepStyle,
+  hiddenLayerStyle,
+  layerStyle,
+  slideInAnimationStyle,
+  slideOutAnimationStyle,
+  stepBoxBottomStyle,
+  stepBoxStyle,
+  stepBoxTopStyle,
+  stepLabelStyle,
+  stepLabelTextStyle,
+  stepStyle,
+  upcomingStepLabelStyle,
+  upcomingStepStyle,
+} from './styles';
+
+type StepCompletionState = 'completed' | 'current' | 'upcoming';
+
+type InternalStepProps = JSX.IntrinsicElements['div'] & {
+  children: React.ReactNode;
+  state: StepCompletionState;
+  stepLabel?: string;
+};
+
+function InternalStep({
+  children,
+  state,
+  stepLabel,
+  className,
+  ...containerProps
+}: InternalStepProps) {
+  const isCurrent = state === 'current';
+  const isUpcoming = state === 'upcoming';
+  const isCompleted = state === 'completed';
+  return (
+    <div
+      className={cx(
+        stepStyle,
+        {
+          [currentStepStyle]: isCurrent,
+          [upcomingStepStyle]: isUpcoming,
+        },
+        className,
+      )}
+      {...containerProps}
+    >
+      {stepLabel && (
+        <div
+          className={cx(stepLabelStyle, {
+            [currentStepLabelStyle]: isCurrent,
+            [completedStepLabelStyle]: isCompleted,
+            [upcomingStepLabelStyle]: isUpcoming,
+          })}
+        >
+          {isCompleted ? (
+            <CheckmarkIcon />
+          ) : (
+            <div className={stepLabelTextStyle}>{stepLabel}</div>
+          )}
+        </div>
+      )}
+      {children}
+      <div className={stepBoxStyle}>
+        <span className={stepBoxTopStyle}>
+          <div />
+        </span>
+        <span className={stepBoxBottomStyle}>
+          <div />
+        </span>
+      </div>
+    </div>
+  );
+}
+
+InternalStep.displayName = 'Step';
+
+type StepItem = InternalStepProps & {
+  step: 'previous' | 'next' | number;
+};
+
+type StepContents = string | number | Array<StepContents>;
+
+type StepElement = React.ReactElement<{ children: StepContents }>;
+export type StepElements = StepElement | Array<StepElement>;
+
+interface StepperProps {
+  children: StepElements;
+  currentStep: number;
+  maxDisplayedSteps: number;
+}
+
+export function Stepper({
+  children,
+  currentStep,
+  maxDisplayedSteps,
+}: StepperProps) {
+  const [visibleLayer, setVisibleLayer] = useState<
+    'previous' | 'next' | 'current'
+  >('current');
+
+  const allSteps: Array<StepItem> = (Array.isArray(children)
```

Comment:
I think I know the answer but just want to double check - wrapping this function inside a `useMemo(fn, [children])` won't give us great value as children will, by default, be a new instance?

Reply:
the function also uses `currentStep`, so the only prop change that shouldn't result in a re-evaluation is `maxDisplaySteps`, but I would generally expect it to not change.